sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/UnionDistributionPushdown.scala

```diff
@@ -0,0 +1,146 @@
+ * 4) the data types of operator's children output attributes are the same
+ */
+case class UnionDistributionPushdown() extends Rule[SparkPlan] {
+
+  def allChildrenSameDataType(unionExec: UnionExec): Boolean = {
+    (Seq(unionExec) ++ unionExec.children).map(_.output).transpose.map(attrs => {
+      val firstAttr = attrs.head
+      attrs.forall(_.dataType == firstAttr.dataType)
+    }).forall(_ == true)
+  }
+
+  def createNewChildrenDistribution(unionExec: UnionExec,
+                                    unionRequired: Distribution): Seq[Distribution] = {
+    unionExec.children.map(unionChild => {
+      // create output attributes map (union output exprId -> child output)
+      val attributes = unionExec.output.map(_.exprId).zip(unionChild.output).toMap
+      unionRequired match {
+        case distribution: ClusteredDistribution =>
+          val newExpressions = distribution.clustering.map(exp => exp.transformUp {
+            case attr: Attribute if attributes.contains(attr.exprId) => attributes(attr.exprId)
+          })
+          ClusteredDistribution(newExpressions, distribution.requireAllClusterKeys,
+            distribution.requiredNumPartitions)
+
+        case distribution => distribution
+      }
+    })
+  }
+
+  override def apply(plan: SparkPlan): SparkPlan = plan.transformUp {
+    case operator: SparkPlan if conf.getConf(SQLConf.UNION_REQUIRED_DISTRIBUTION_PUSHDOWN) &&
+      operator.children.exists(_.isInstanceOf[UnionExec]) =>
+      val newChildren = operator.children.zipWithIndex.map {
+        case (unionExec: UnionExec, childIndex: Int) if !unionExec.outputPartitioning.
+          satisfies(operator.requiredChildDistribution(childIndex)) =>
+          val unionRequired = operator.requiredChildDistribution(childIndex)
+          unionRequired match {
+            case _: ClusteredDistribution if allChildrenSameDataType(unionExec) =>
+              val requiredChildDistribution: Seq[Distribution] =
+                createNewChildrenDistribution(unionExec, unionRequired)
+              val numPartitions = unionRequired.requiredNumPartitions
+                .getOrElse(conf.numShufflePartitions)
+              val outputPartitioning = unionRequired.createPartitioning(numPartitions)
+              UnionZipExec(unionExec.children, requiredChildDistribution,
+                outputPartitioning)
+
+            case _: Distribution => unionExec
+          }
+
+        case (child, _) => child
+      }
+      operator.withNewChildren(newChildren)
+  }
+}
+
+/**
+ * Class representing partitions of UnionZipRDD, which maintains the list of
+ * corresponding partitions of parent RDDs.
+ */
+private[spark]
+class UnionZipRDDPartition(@transient val rdds: Seq[RDD[_]],
+                           override val index: Int) extends Partition {
+  var parents: Array[Partition] = rdds.map(_.partitions(index)).toArray
+
+  override def hashCode(): Int = index
+
+  override def equals(other: Any): Boolean = super.equals(other)
```

Comment:
Is `equals` required? Doesn't do much.

Reply:
Thanks for your advice ! I have updated it.

Reply:
Thanks for your advice ! I have updated it.